### ISSUE TITLE
Clarify behaviour of NtCreateFile with OBJ_CASE_INSENSITIVE and setCaseSensitiveInfo 

### ISSUE THREAD

**Issue by Bigpet** [Microsoft/WSL#3359]
### Reproduction steps

First, the steps to reproduce this easily when having visual studio installed are as follows (if you don't have visual studio installed you can use [this small program](https://gist.github.com/Bigpet/7b9740b8af8a8aa7ed60746654f71a6a) to reproduce the failure):

    mkdir C:\T  
    mkdir C:\T\P  
    cd C:\T\P  
    echo "int main(){}" > test.cpp  
    fsutil.exe file setCaseSensitiveInfo ../ enable  
    cl test.cpp  

This will make cl.exe not find the file test.cpp .

### Details of the specific failure mode

cl.exe retrieves the full path of the the file. Then it transforms it to all lowercase using `LCMapStringW` with `LCMAP_LOWERCASE`. Finally it then passes the resulting string `\??\c:\t\p` to `NtCreateFile` with `OBJ_CASE_INSENSITIVE` set. The call then returns `STATUS_OBJECT_NAME_NOT_FOUND` and cl.exe concludes that test.cpp does not exist.

### Question

Now on to the actual question:

Is it desired that `NtCreateFile` with `FILE_OPEN` and `OBJ_CASE_INSENSITIVE` return `STATUS_OBJECT_NAME_NOT_FOUND` when CaseSensitiveInfo is enabled?  
If so, can we get the specific circumstances officially documented?  
If not, which file should `NtCreateFile` open if multiple candidates are present?  For example if there are two folders `C:\Test123\P` and `C:\tEst123\P` and someone calls `NtCreateFile` with `C:\teSt123\P`, which one should be the one that's opened?

Related links:  
- https://github.com/Microsoft/WSL/issues/2937 -  related issue, but it's just one manifestation of the core-issue  
- https://developercommunity.visualstudio.com/content/problem/287514/clexe-cant-find-files-in-casesensitive-folders.html - the bug I filed for the specific cl.exe case
- https://blogs.msdn.microsoft.com/commandline/2018/02/28/per-directory-case-sensitivity-and-wsl/ - Blog post about the pre-directory case sensitivity
- https://docs.microsoft.com/en-us/windows/desktop/api/winternl/nf-winternl-ntcreatefile - NtCreateFile reference

**Comment by Bigpet** [comment]
@onomatopellan True, it was mentioned in #2937 as well. That would remove most of the cases of accidentally stumbling across this.

The possibility to enable the flag however remains. So I would still like this to be clarified since the current publicly available documentation seems insufficient.

**Comment by therealkenc** [comment]
>  So I would still like this to be clarified since the current publicly available documentation seems insufficient.

Your "actual questions" posed are largely rhetorical. The answer to your first question is (rhetorically) "yes" (although a better adverb would be "intended" rather than "desired"). The "if not..." qualification in your second question means the question doesn't apply, because "not". The [`NtCreateFile()`](https://docs.microsoft.com/en-us/windows/desktop/api/winternl/nf-winternl-ntcreatefile) docs should be updated to indicate that the `OBJ_CASE_INSENSITIVE` flag is overridden on case sensitive directories, sure, absolutely. But such an update won't tell you something you don't already know because you've seen the intended behaviour yourself. The "specific circumstances" under which the `OBJ_CASE_INSENSITIVE` flag is overridden is (wait for it...) when the directory has the case sensitivity flag set. That's rhetorical too. You can check whether the flag is set on a given directory with `fsutil.exe`.

None of the above of which is meant as an endorsement of, nor opposition to, the behaviour. All I am saying is the behaviour doesn't require a lot of clarification. It is crystal clear.
